### DISCUSSION TITLE
Micropython compatibility

### DISCUSSION THREAD

**Discussion by Sylvain-BROCAS** [falconry/falcon#2418] [General]
Hello,
Thank you for this great module. I'm currently working on an embedded device running micropython on an ESP32.
I'm wondering if falcon is compatible with micropython as it's supported by Cpython.

Thank you for your attention

**Comment by vytas7** [comment]
Hi @Sylvain-BROCAS!
Honestly, I don't know. I think it would be easiest to simply test it, if you have a suitable device or emulator :slightly_smiling_face: 

Falcon has no hard dependencies, but it still needs a WSGI or ASGI server to run. On normal CPython or PyPy, you can use `wsgiref.simple_server` in a pinch, but as I understand it is unavailable in Micropython.

You could try installing Falcon on your device, and also installing [adafruit-circuitpython-wsgi](https://pypi.org/project/adafruit-circuitpython-wsgi/).

Then you could try running it along the lines of this example: https://github.com/adafruit/Adafruit_CircuitPython_WSGI/blob/main/examples/wsgi_simpletest.py:
```python
import falcon
from adafruit_wsgi.esp32spi_wsgiserver import WSGIServer


class HelloResource:
    def on_get(self, req, resp):
        resp.media = {'message': 'Hello, World!'}


app = falcon.App()
app.add_route('/hello', HelloResource())

wsgi_server = WSGIServer(8000, application=app)

# print('open this IP in your browser: ', esp.pretty_ip(esp.ip_address))

wsgi_server.start()
while True:
    # Our main loop where we have the server poll for incoming requests
    try:
        wsgi_server.update_poll()
        # Could do any other background tasks here, like reading sensors
    except (ValueError, RuntimeError) as e:
        print("Failed to update server, restarting ESP32\n", e)
        # Need to import/initialize this `wifi`...
        # wifi.reset()
        continue
```
(totally untested)

Let us know if you get it to work!

**Comment by Sylvain-BROCAS** [comment]
OK, I can't test right now, but I'll do and share the result as soon as possible

**Comment by Sylvain-BROCAS** [comment]
In the first version of my program, I used microdot ([https://github.com/miguelgrinberg/microdot](https://github.com/miguelgrinberg/microdot)) that seems to run as asgi. I runs well on my ESP32.
I'I do some tests to use falcon with microdot and give my conclusions here then.

**Comment by CaselIT** [comment]
moving to discussion, since there isn't any issue here

**Comment by vytas7** [comment]
Microdot doesn't run as ASGI though, it _can_ run as ASGI if you have an ASGI server, but unfortunately it doesn't seem to expose its built-in server for general use.

**Comment by vytas7** [comment]
Hi again @Sylvain-BROCAS,
I have now tried the `micropython/unix` Docker container, and unfortunately I am 100% sure Falcon is incompatible with Micropython :imp:

The framework relies on a fair number of stdlib modules that are unavailable:
* `__future__`
* `dataclasses`
* `datetime`
* `enum`
* `inspect`
* `logging`
* `string`
* `typing`
* `warnings`
...and probably a few more.

Some of these like `typing` can probably be shimmed away relatively easily, but some others would require more effort to replace, not sure if that is even easily possible.

**Reply by Sylvain-BROCAS** [reply]
OK, thank you for your time and attention.
It's good to know.

**Reply by vytas7** [reply]
It would be cool to have a flavour of Falcon (automatically stripped down from the main version) running on Micropython.
But I think the amount of available RAM might be another issue... even though Falcon uses less than many other frameworks, but from what I've seen many ESP32 variants only have 520K, where even less is available to Micropython.
How much RAM do you have to play with @Sylvain-BROCAS?

**Reply by Sylvain-BROCAS** [reply]
I give you back the information as soon as I can.

I may use microdot instead that can do the same thing for my simple application.

**Reply by vytas7** [reply]
Yes, do go ahead with Microdot, it is unlikely Falcon is going to support Micropython anytime soon, if ever. It would be a cool project though, and I'm just trying to understand how feasible this is.

**Reply by Sylvain-BROCAS** [reply]
Hi,
Here is what I found for ma board :
![image](https://github.com/user-attachments/assets/0cf3ab3f-7eb7-41f8-a231-33e5785664f8)

I read memory information on the board i'm currently using : 
![image](https://github.com/user-attachments/assets/d787eb8e-b50e-43e1-9a39-6223490b7036)
